M2/Macaulay2/packages/Varieties/SheafMaps.m2

```diff
@@ -0,0 +1,890 @@
+    phi := matrix shphi;
+    M := module source shphi;
+    eta := inducedMap(truncate(d, M), source phi);
+    isLiftable(phi, eta))
+
+--if phi is in the image of Hom(eta,target phi), this code
+--computes the actual lift
+lift' = method()
+lift'(Matrix, Matrix) := Matrix => (phi, eta) -> (
+    newPhi := homomorphism' phi;
+    homomorphism(newPhi // Hom(eta, target phi))
+    )
+
+--lifts a sheaf map shphi represented by a module map
+--phi : M(\geq d) --> N to a map M(\geq e) --> N that represents
+--the same morphism of sheaves, if possible
+--WARNING: this method does not actually verify if the lift is possible
+lift'(SheafMap,ZZ) := SheafMap => (shphi,e) -> (
+    d := degree shphi;
+    phi := matrix shphi;
+    M := module source shphi;
+    eta := inducedMap(truncate(e,M),source phi);
+    sheafMap(lift'(phi, eta), e))
+
+--lifts a sheaf map shphi represented by a module map
+--phi : M(\geq d) --> N to a map M(\geq e) --> N that represents
+--the same morphism of sheaves, for the smallest possible value of e
+--WARNING: this method does not actually verify if the lift is possible
+lift SheafMap := SheafMap => o -> shphi -> (
+    d := degree shphi;
+    M := module source shphi;
+    m := min flatten degrees M;
+    while isLiftable(shphi,d-1) and d > m do d = d-1;
+    lift'(shphi, d))
+
+-*lift(Matrix,Matrix) := Matrix => opts -> (phi,eta) -> (
+    newPhi := homomorphism'(phi);
+    newPhi//Hom(eta,target phi)
+    )*-
+
+-- TODO: this needs to be improved: there are more inducedMap methods to add
+inducedMap(CoherentSheaf, CoherentSheaf) := SheafMap => opts -> (G, F) -> map(G, F, inducedMap(module G, module F, opts))
+-- TODO: what should this operation be?
+-- inducedMap(CoherentSheaf, CoherentSheaf, SheafMap)
+
+-----------------------------------------------------------------------------
+-- Direct sums and components
+-----------------------------------------------------------------------------
+--WARNING: the current direct sum of sheaves does not cache components
+SheafMap.directSum = args -> (
+    assert(#args>0);
+    X := variety args#0;
+    if not same apply(args, variety) then error "expected maps of sheaves over the same variety";
+    DS := map(
+	directSum apply(args, target),
+	directSum apply(args, source),
+	directSum apply(args, matrix));
+    DS.cache.components = toList args;
+    DS)
+
+directSum SheafMap := SheafMap => phi -> directSum(1:phi)
+SheafMap ++ SheafMap := SheafMap => (phi, psi) -> directSum(phi, psi)
+
+components SheafMap := List => phi -> if phi.cache.?components then phi.cache.components else {phi}
+
+-----------------------------------------------------------------------------
+-- Tensors
+-----------------------------------------------------------------------------
+-- TODO: take care of the case when the rings are different
+tensor(SheafMap, SheafMap) := SheafMap => (phi, psi) -> (
+    map(target phi ** target psi,
+	source phi ** source psi,
+	matrix phi ** matrix psi))
+
```

Comment:
Does one really not need to consider the truncation degrees when forming tensor products of ```SheafMaps```?